c++/src/TypeImpl.cc

```diff
@@ -121,6 +121,32 @@ namespace orc {
     return scale;
   }
 
+  Type& TypeImpl::setAttribute(const std::string& key,
+                     const std::string& value) {
+    attributes[key] = value;
+    return *this;
+  }
+
+  Type& TypeImpl::removeAttribute(const std::string& key) {
+    attributes.erase(key);
+    return *this;
+  }
+
+  std::vector<std::string> TypeImpl::getAttributeKeys() const {
+    std::vector<std::string> ret;
+    ret.reserve(attributes.size());
+    for (auto& attribute : attributes) {
+      ret.push_back(attribute.first);
+    }
+    return ret;
+  }
+
+  std::string TypeImpl::getAttributeValue(const std::string& key) const {
```

Comment:
I'd prefer to add a new interface`bool hasAttributeKey(const std::string& key) const` to check existence and throw here if the key does not exist.

Reply:
I've added hasAttributeKey() to Type, and throw std::range_error in removeAttribute() and getAttributeValue() when key is not present